sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Cast.java

```diff
@@ -0,0 +1,582 @@
+      }
+
+      return outputMap;
+    } else {
+      if (inputFieldType.getTypeName().equals(outputFieldType.getTypeName())) {
+        return inputValue;
+      } else {
+        KV<TypeName, TypeName> key =
+            KV.of(inputFieldType.getTypeName(), outputFieldType.getTypeName());
+        return Inference.ALL_MAP.get(key).apply(inputValue);
+      }
+    }
+  }
+
+  /** Configures casting for nullable fields. */
+  public enum Nullability {
+    /** Can cast only if nullability is the same. */
+    SAME,
+    /** Can cast non-nullable fields to nullable. */
+    WEAKEN,
+    /** Can cast nullable fields to non-nullable and forth. */
+    IGNORE
+  }
+
+  /** Configures casting of primitive types. */
+  public enum Type {
+    /** Can cast only if types are the same. */
+    SAME,
+    /** Can cast only if an output type is a supertype of an input type. */
+    WIDEN,
+    /** Can cast if an output type is a supertype or subtype of an input type. */
+    NARROW
+  }
+
+  /** Configures casting of rows. */
+  public enum Shape {
+    /** Can cast rows only if field names are the same. */
+    SAME,
+    /** Can cast rows if output field names is subset of input field names. */
+    PROJECTION
+  }
+
+  /** Describes compatibility errors during casting. */
+  @AutoValue
+  public abstract static class CompatibilityError {
+
+    public abstract List<String> path();
+
+    public abstract String message();
+
+    public static CompatibilityError create(List<String> path, String message) {
+      return new AutoValue_Cast_CompatibilityError(path, message);
+    }
+  }
+
+  static class Inference {
+
+    @FunctionalInterface
+    interface Converter {
+
+      Object apply(Object value);
+    }
+
+    @AutoValue
+    abstract static class Result<T> {
+
+      abstract List<String> path();
+
+      abstract Optional<T> value();
+
+      abstract List<CompatibilityError> errors();
+
+      static <T> Result<T> create(
+          List<String> path, Optional<T> value, List<CompatibilityError> errors) {
+        return new AutoValue_Cast_Inference_Result<>(path, value, errors);
+      }
+
+      public static <T> Result<T> of(List<String> path, T value) {
+        return create(path, Optional.of(value), Collections.emptyList());
+      }
+
+      public static <T> Result<T> error(List<String> path, String message) {
+        return Result.create(
+            path,
+            Optional.empty(),
+            Collections.singletonList(CompatibilityError.create(path, message)));
+      }
+    }
+
+    static final List<Nullability> NULLABILITY_ORDER =
+        ImmutableList.of(Nullability.SAME, Nullability.WEAKEN, Nullability.IGNORE);
+
+    static final List<Shape> SHAPE_ORDER = ImmutableList.of(Shape.SAME, Shape.PROJECTION);
+
+    static final List<Type> TYPE_ORDER = ImmutableList.of(Type.SAME, Type.WIDEN, Type.NARROW);
+
+    // TODO extend this list
+
+    static final Map<KV<TypeName, TypeName>, Converter> TYPE_WIDEN =
```

Comment:
The set of types is not dynamic, so I would suggest just using a nested `switch` instead of a map for what to do for each type constructor.

Reply:
Keys are used for checking if there is a conversion between two types